master/internal/rm/kubernetesrm/pods.go

```diff
@@ -1275,3 +1299,50 @@ func (p *pods) listConfigMapsInAllNamespaces(
 
 	return res, nil
 }
+
+func extractTCDs(resourcePoolConfigs []config.ResourcePoolConfig) map[string]*model.TaskContainerDefaultsConfig {
+	result := map[string]*model.TaskContainerDefaultsConfig{}
+
+	for _, config := range resourcePoolConfigs {
+		result[config.PoolName] = config.TaskContainerDefaults
+	}
+
+	return result
+}
+
+func taintTolerated(taint k8sV1.Taint, tolerations []k8sV1.Toleration) bool {
+	for _, toleration := range tolerations {
+		if toleration.ToleratesTaint(&taint) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func allTaintsTolerated(taints []k8sV1.Taint, tolerations []k8sV1.Toleration) bool {
+	for _, taint := range taints {
+		if !taintTolerated(taint, tolerations) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func extractSlotInfo(node model.AgentSummary) (numSlots int, devType device.Type) {
+	var gpuSlots, cpuSlots int
+	for _, slot := range node.Slots {
+		if slot.Device.Type == device.CPU {
+			cpuSlots++
+		} else if slot.Device.Type == device.CUDA {
+			gpuSlots++
+		}
+	}
+
+	if gpuSlots > 0 {
```

Comment:
non-blocking: Maybe it would be more robust to switch based on whether `slotType = cpu`. The edge case is if you launch a gpu node but have cpu slots configured. A huge waste of money but maybe 🤷

Reply:
Not sure I understand. Do you mean if CPU slots > 0 return that rather than preferring GPU slots?

Reply:
Because in `values.yaml` we have a switch for whether it's CPU or GPU training. If it's set to CPU training, we would be measuring CPU slots in all cases, even if there are GPU devices on the node.